airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py

```diff
@@ -0,0 +1,78 @@
+#
+# Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+#
+
+from typing import Any, Dict, List, Optional, Union
+
+from airbyte_cdk.models import AirbyteRecordMessage
+from genson import SchemaBuilder
+from genson.schema.strategies.object import Object
+
+
+class NoRequiredObj(Object):
+    """
+    This class has Object behaviour, but it does not generate "required[]" fields
+    every time it parses object. So we dont add unnecessary extra field.
+    """
+
+    def to_schema(self):
+        schema = super(Object, self).to_schema()
+        schema["type"] = "object"
+        if self._properties:
+            schema["properties"] = self._properties_to_schema(self._properties)
+        if self._pattern_properties:
+            schema["patternProperties"] = self._properties_to_schema(self._pattern_properties)
+        return schema
+
+
+class NoRequiredSchemaBuilder(SchemaBuilder):
+    EXTRA_STRATEGIES = (NoRequiredObj,)
+
+# This type is inferred from the genson lib, but there is no alias provided for it - creating it here for type safety
+InferredSchema = Dict[str, Union[str, Any, List, List[Dict[str, Union[Any, List]]]]]
+
+
+class SchemaInferrer:
+    """
+    This class is used to infer a JSON schema which fits all the records passed into it
+    throughout its lifecycle via the accumulate method.
+
+    Instances of this class are stateful, meaning they build their inferred schemas
+    from every record passed into the accumulate method.
+
+    """
+
+    builders: Dict[str, SchemaBuilder]
+
+    def __init__(self):
+        self.builders = {}
+
+    def accumulate(self, record: AirbyteRecordMessage):
+        """Uses the input record to add to the inferred schemas maintained by this object"""
+        stream_name = record.stream
+        builder = None
+        if stream_name not in self.builders:
+            builder = NoRequiredSchemaBuilder()
+            self.builders[stream_name] = builder
+        else:
+            builder = self.builders[stream_name]
+        builder.add_object(record.data)
+
+    def get_inferred_schemas(self) -> Dict[str, InferredSchema]:
+        """
+        Returns the JSON schemas for all encountered streams inferred by inspecting all records
+        passed via the accumulate method
+        """
+        schemas = {}
+        for stream_name, builder in self.builders.items():
+            schemas[stream_name] = builder.to_schema()
+        return schemas
+
+    def get_stream_schema(self, stream_name: str) -> Optional[InferredSchema]:
```

Comment:
The original tech spec outlined the `accumulate` and `get_inferred_schemas` functions, but for context what's the use case for a single stream or how are we using this `get_stream_schema()` function
Also why do we no longer need the `reset()`?

Reply:
fwiw when writing the spec i didn't account for this object needing to infer schema on multiple streams (that iface assumed a single stream). I think this iface makes more sense.

Reply:
Yeah, when integrating with the API I noticed that for that use case we only need a specific one and this API seemed nicer. The `get_inferred_schemas` will still be nice for the planned integration with the `read` command / implementing a separate command for schema inferrence

Reply:
I dropped the `reset` because in the places where we plan to use this class it's much easier to just lose the reference to the inferrer instance and have the runtime clean it up, then create a new instance for the next usage. Happy to change that, but it doesn't seem heavy enough to explicitly keep an instance or a pool of them around.